MHUB/IDC - Implementing the Prostate158 whole prostate gland segmentation model available in MONAI (T2 only)

ccosmin97 opened the pull request:
Adding template for PR request

ccosmin97 committed:
init commit for prostate158 model

ccosmin97 committed:
added meta.json, config and dockerfiles, tried testing with and without center cropping

ccosmin97 committed:
fixed dockerfile

ccosmin97 commented:
I finished the implementation of the prostate158 model in mhub.ai configuration. Generated outputs are conform with my study. 

Instructions for testing can be found [here -- .md file](https://github.com/MHubAI/models/files/13708059/Overview.md).
Input data is here -- [test_data.zip](https://github.com/MHubAI/models/files/13708224/test_data.zip) and output data -- [monai_prostate158.zip](https://github.com/MHubAI/models/files/13708226/monai_prostate158.zip) is here. Testing sample size is one serieUID only -- although that can be changed easily if deemed necessary for testing.


@kbressem The meta information for the model is located [here](https://github.com/ccosmin97/models/blob/monai_prostate158/models/monai_prostate158/meta.json). The default [workflow](https://github.com/ccosmin97/models/blob/monai_prostate158/models/monai_prostate158/config/default.yml) applies the same center-cropping strategy as defined in the prostate158 model training process. For DICOM SEG conversion purposes, the output segmentation file is padded back to the original (pre-centering) image size. This center-cropping strategy is a parameter of the [Prostate158Runner](https://github.com/ccosmin97/models/blob/monai_prostate158/models/monai_prostate158/utils/Prostate158Runner.py), and can be changed to False.

@LennyN95 Please let me know whenever you have a minute if my testing procedure is enough or needs to be extended/modified, thank you!

kbressem commented:
The meta information seems all correct to me.

LennyN95 commented on models/monai_prostate158/dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+ENV BUNDLE_ROOT=/app/models/monai_prostate158/bundle/prostate_mri_anatomy
+
+# specify nnunet specific environment variables
+# ENV WEIGHTS_FOLDER=$WEIGHTS_DIR
```
Remove lines

LennyN95 commented on models/monai_prostate158/utils/Prostate158Runner.py:
```diff
@@ -0,0 +1,85 @@
+import json
+
+@IO.Config('apply_center_crop', bool, False, the='flag to apply center cropping to input_image')
+
```
I suggest we remove all empty lines between decorators and their target. This is a personal preference but it makes the relation more clear imo.

LennyN95 commented on models/monai_prostate158/utils/Prostate158Runner.py:
```diff
@@ -0,0 +1,85 @@
+
+@IO.Config('apply_center_crop', bool, False, the='flag to apply center cropping to input_image')
+
+class Prostate158Runner(ModelRunner):
```
Please inherit from `Module` directly. I'll subsequently remove all abstract module classes such as the ModelRunner in the newr future. This is to add transparency as theyre no longer required.

LennyN95 commented on models/monai_prostate158/utils/Prostate158Runner.py:
```diff
@@ -0,0 +1,85 @@
+        out_path = glob.glob(os.path.join(out_dir, "**", 
+        "*.nii.gz"), recursive=True)[0]
+
+        if self.apply_center_crop:
```
Why do we have a script (`center_crop.py`)for preprocessing and a custom implementation for postprocessing?

ccosmin97 replied:
center_crop.py comes from the MONAI bundle when building the docker image, whereas the postprocessing is done by me only, necessary to pad the resulting seg output to the original image space, right before conversion to SEG DICOM.  To make it 'cleaner', the only way I see is to take out the center_crop.py out of the monai bundle and add it to mhub as a utility script, alongside the postprocessing (line 65+). Not sure though if it is much better. 

FYI regarding the post-processing : Another strategy would have been to use the cropped image as a reference and then no need for this post-processing code, however I personally did not like it since it modifies the original image and padding with 0 the resulting seg image is not affecting it.

LennyN95 replied:
Thx for the explaination. That's just fine then.

LennyN95 commented on models/monai_prostate158/utils/Prostate158Runner.py:
```diff
@@ -0,0 +1,85 @@
+
+        # bring input data in nnunet specific format
+        # NOTE: only for nifti data as we hardcode the nnunet-formatted-filename (and extension) for now.
+        assert in_data.type.ftype == FileType.NIFTI
```
These don't do any harm, however, as you explicitly request `nifti`from the `@IO.Input` decorator you can be sure to only get files of type NIFTI here. The extension could be different though (e.g. uncompressed nifti).

LennyN95 commented on models/monai_prostate158/utils/Prostate158Runner.py:
```diff
@@ -0,0 +1,85 @@
+    @IO.Instance()
+    @IO.Input("in_data", 'nifti:mod=mr', the="input T2 sequence data to run prostate158 on")
+    @IO.Output('out_data', 'monai_prostate158.nii.gz', 
+               'nifti:mod=seg:model=Prostate158:roi=PROSTATE_TRANSITION_ZONE,PROSTATE_PERIPHERAL_ZONE', 
```
Maybe add monai to the model name?

LennyN95 commented on models/monai_prostate158/utils/Prostate158Runner.py:
```diff
@@ -0,0 +1,85 @@
+from mhubio.modules.runner.ModelRunner import ModelRunner
+import json
+
+@IO.Config('apply_center_crop', bool, False, the='flag to apply center cropping to input_image')
```
Do we have explained (e.g. on the model card) when and why you want to apply center crop? I cannot find any notion when to disable this setting in the intended use. If the flag isn't actually optional we shouldn't expose it. If it's optional, we should give a hint of twhen to disable it ;)

ccosmin97 replied:
center_crop should be applied by default, since it is recommended by prostate158 for both training and inference.

Their center_crop strategy is enabled by default (called by monai run), but the user can disable it by passing center_crop=False in the .yml file.

LennyN95 replied:
> center_crop should be applied by default, since it is recommended by prostate158 for both training and inference.

But the default value is `False`;)

Pls change to `@IO.Config('apply_center_crop', bool, True, the='flag to apply center cropping to input_image')`

LennyN95 reviewed (changes requested):
Looking good overall. Some small issues to be adressed.

ccosmin97 committed:
Merge branch 'MHubAI:main' into monai_prostate158

ccosmin97 committed:
Update Dockerfile

removed commented code

ccosmin97 committed:
Update default.yml

delete center_crop = True argument, since its enabled by default

renamed model_name in DsegConverter

ccosmin97 committed:
Update Prostate158Runner.py

ccosmin97 committed:
Update meta.json

added external evaluation tables and references

ccosmin97 reviewed (commented):
reviewed changes.

LennyN95 commented on models/monai_prostate158/config/default.yml:
```diff
@@ -0,0 +1,36 @@
+    import_dir: sorted_data
+    sort_data: true
+    meta: 
+      mod: 'mr'
```
`mod: '%Modality'`

LennyN95 commented on models/monai_prostate158/config/default.yml:
```diff
@@ -0,0 +1,36 @@
+
+  DataOrganizer:
+    targets:
+    - dicomseg-->[i:sid]/prostate158.seg.dcm
```
for consistency: `<model_name>.seg.dcm` where model_name is the name of the folder and also the name used in meta.json.

LennyN95 commented on models/monai_prostate158/dockerfiles/Dockerfile:
Nice & clean!!

LennyN95 commented on models/monai_prostate158/meta.json:
```diff
@@ -0,0 +1,131 @@
+        "devteam": "Lisa C. Adams, Keno K. Bressem",
+        "type": "Prostate158 (U-Net structure for prostate segmentation)",
+        "date": {
+            "weights": "2022",
```
Do we have more exact dates?

ccosmin97 replied:
yes I added the month information.

LennyN95 commented on models/monai_prostate158/meta.json:
```diff
@@ -0,0 +1,131 @@
+        "cite": "Lisa C. Adams and Marcus R. Makowski and Günther Engel and Maximilian Rattunde and Felix Busch and Patrick Asbach and Stefan M. Niehues and Shankeeth Vinayahalingam and Bram {van Ginneken} and Geert Litjens and Keno K. Bressem, Prostate158 - An expert-annotated 3T MRI dataset and algorithm for prostate cancer detection",
+        "license": {
+            "code": "Apache 2.0",
+            "weights": ""
```
licence?

ccosmin97 replied:
added.

ccosmin97 committed:
Update default.yml

changed seg output name and added '%Modality' filter

ccosmin97 committed:
Update meta.json

added license information

ccosmin97 reviewed (commented):
reviewed changes and added/modified model code based on @LennyN95 suggestions.

LennyN95 commented:
Looking good. @ccosmin97 shall we proceed with testing? If you feel ready, you can create a /test comment as described in our documentation [here](https://github.com/MHubAI/documentation/blob/main/documentation/mhub_contribution/testing_phase.md).

ccosmin97 commented:
> Looking good. @ccosmin97 shall we proceed with testing? If you feel ready, you can create a /test comment as described in our documentation [here](https://github.com/MHubAI/documentation/blob/main/documentation/mhub_contribution/testing_phase.md).

Ok thanks @LennyN95 I will provide testing results later today.

ccosmin97 commented:
/test

```yaml
sample:
  idc_version: 17.0
  data:
  - SeriesInstanceUID: 1.3.6.1.4.1.14519.5.2.1.7311.5101.117787454837907188090392680606
    aws_url: s3://idc-open-data/4a15953b-3e16-4aa3-beb3-53187d84b6ed/*
    path: dicom

reference:
  url: https://www.dropbox.com/scl/fi/j6ouks1vqfeofaek8n3dk/output.zip?rlkey=y0p4b9pi2ot099iyxcjiqpdoj&dl=0
```

---

<details><summary>Test Results (24.03.06_09.47.47_ROPNerVqB9)</summary>

```yaml
id: a09b9f72-406f-47dd-bd38-ab883380bc8b
date: '2024-03-06 09:51:53'
checked_files:
- file: monai_prostate158.seg.dcm
  path: /app/test/src/1.3.6.1.4.1.14519.5.2.1.7311.5101.117787454837907188090392680606/monai_prostate158.seg.dcm
  checks:
  - checker: DicomsegContentCheck
    notes:
    - label: Segment Count
      description: The number of segments identified in the inspected dicomseg file.
      info: 2
summary:
  files_missing: 0
  files_extra: 0
  checks:
    DicomsegContentCheck:
      files: 1
conclusion: true
```

</details>

ccosmin97 commented:
@LennyN95 

FYI
trying to run the mhub prostate158_model as suggested here
https://github.com/MHubAI/documentation/blob/main/documentation/mhub_contribution/testing_phase.md
by:
`MHUB_OUTPUT_DIR=/path/to/your/output/folder`
`docker run dev/$MHUB_MODEL_NAME:latest -v /path/to/your/sample/data:/app/data/input_data:ro -v $MHUB_OUTPUT_DIR:/app/data/output_data `
**fails.**

However, running:
`docker run -it --rm --network=none -v /home/exouser/Documents/mhub_exps/prostate_158/idc_data/dicom_data/subject_1:/app/data/input_data:ro -v /home/exouser/Documents/mhub_exps/prostate_158/mhub_out/out_subject_1:/app/data/output_data  --entrypoint bash dev/monai_prostate158:latest`
then
`mhub.run default`
produces the correct output, i.e the one I tested and documented above.

Any idea why this is happening? A mistake on my end?

LennyN95 commented:
@ccosmin97 The docker-image needs to be declared after all docker run arguments. Any argument after the image name will be passed to the entry point (`mhub.run` by default).

The syntax is: `docker run [docker run args] mhubai/$model_name:latest [mhub.run arrgs]` where `[docker run args]` usually are the input and output volume mounts (`-v /local/abs/path/:/app/data/input_data`) and `--rm -t --gpus all` to remove the container after execution, stream output to the console and make all GPUs accessible to the container. `[mhub.run arrgs]`  then can be, e.g., `--workflow default` to specify a workflow or `--debug` to enable debug mode (all available options are shown when passing `--help`). 

**Note that when performing a test run, there must be no mhub.run argument set!**
